---
title: "Transcriptome Assembly and Evaluation"
subtitle: "using Trinity and Transrate"
author: "Nate Olson & Leann Biancani"
date: "May 15, 2016"
output: [github_document, pdf_document]
bibliography: tutorial.bib 
---
```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
library(knitr)
library(readr)
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(ggplot2)
library(GGally)
```

```{r source_dir, warning=FALSE, message=FALSE, echo = FALSE}
results_list <- list(
      no_mods = "data/no_mods/Trinity_sequences.Trinity.fixed/",
      corrected = "data/corrected/Assembly4_Corrected.Trinity.fixed/",
      trimmed = "data/trimmed/Assembly2_Trimmed.Trinity.fixed/",
      trimmedcorrected = "data/trimmedcorrected/Assembly3_TrimmedCorrected.Trinity.fixed/"
)
```

## Background 

#### Transcriptomics
Transcriptomics is the study of the complete set of RNA transcripts in a given sample. Various aims of transcriptomic studies include cataloging all species of transcripts (mRNA, non-coding RNAs, smRNAs), determining gene structure, studying splicing patterns, identifying post-transcriptional modifications, or investigating changes in expression patterns [@Wang:2009di]. Most RNA sequencing (RNA-seq) methods utilize short-read sequencing technologies to sequence a library of cDNA fragments generated from an isolation of RNA molecules. In order to characterize transcripts, these short reads sequences must be assembled into contiguous sequences (contigs). To assemble a transcriptome, reads may either be aligned to a known reference genome or transcriptome sequence, or assembled *de novo*.

#### De novo Transcriptome assembly
*De novo* transcriptome assembly algorithms build contigs from RNA-seq data without the use of a reference genome. These assemblies, however, can be highly variable. There are many different algorithms available and a variety of parameters that can be specified for each method, resulting in the ability to generate multiple assemblies from a single set of reads [@smith2015transrate]. As such, it is necessary to be able to evaluate the quality and accuracy of a transcriptome assembly.
  
Trinity [@Grabherr:2011jb, @haas2013novo]
            
#### Transcriptome assembly evaluation  
Two basic methods exist for evaluating transcriptomes. Reference based methods compare the assembly to known genome or transcriptome sequences, while read based methods map the RNA-seq reads back to the assembly.

## Transcriptome Assembly and Evaluation

#### Source data  
A single set of raw RNA-seq reads for _C. elegans_ was identified and downloaded from the SRA database.

_Command Line_
```
Fastq-dump -I --split-files SRR2969230
```

__Dataset Information:__

* Accession number: SRR2969230
* Description:

>
Total RNA from tissue disrupted “whole worm” samples was isolated and amplified using the NuGen Ovation RNAseq v2 kit. Amplified cDNA was sheared to 200bp fragments. Library was prepared using the Illumina TruSeq DNA Sample Prep kit.

* Study: Caenorhabditis elegans Transcriptome or Gene expression BioProject

> 
Abstract: Insulin/IGF-1 signaling (IIS) is a critical regulator of the most important biological decisions, from rates of growth, development, and metabolism, to reproduction and longevity. However, the neuron-specific IIS/FOXO targets that regulate these functions are still unknown. By isolating C. elegans neurons for transcriptional profiling, we identified both the wild-type and IIS/FOXO neuronal transcriptomes. Here we deposit the sequences for wt adult neurons and IIS mutants, as well as for isolated adult mechanosensory neurons.

#### Relevant _C. elegans_ Reference Sequences  

* Transcriptome: ftp://ftp.ensemblgenomes.org/pub/release-31/metazoa/gtf/caenorhabditis_elegans/Caenorhabditis_elegans.WBcel235.31.gtf.gz   
* Genome: ftp://ftp.ensemblgenomes.org/pub/release-31/metazoa/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.31.dna.chromosome.*.fa.gz   
* Proteome: ftp://ftp.ensemblgenomes.org/pub/release-31/metazoa/fasta/caenorhabditis_elegans/pep/Caenorhabditis_elegans.WBcel235.31.pep.all.fa.gz    

#### Data Processing

##### Read Trimming

RNA-seq reads were trimmed with [Trim Galore v0.4.0](http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/). The user guide for this version is available [here](http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/trim_galore_User_Guide_v0.4.0.pdf). Trim Galore is a wrapper script that automates adapter trimming and quality control using the programs [Cutadapt](https://cutadapt.readthedocs.io/en/stable/) and [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

Using the following commands, Trim Galore removes low quality base calls from the 3' end using a Phred score threshold of 20, finds and removes adapter sequences from the 3' end using standard Illumina paired-end adapter sequences, clips 5 bp from the 5' end to combat bias due to to the end-repair libary prep step, clips an additional 5 bp from the 3' end to further improve quality, and discards any reads that become shorter than 20 bp due to either quality filtering or trimming.

_Command Line_
```
trim_galore -q 30 \
  --paired \
  --fastqc \
  --clip_R1 5 --clip_R2 5 \
  --three_prime_clip_R1 5 --three_prime_clip_R2 5 \
  SRR2969230_1.fastq SRR2969230_2.fastq
```

##### Error Correction

Sequencing errors in the RNA-seq reads were corrected using [BayesHammer](http://bioinf.spbau.ru/en/spades/bayeshammer) which is part of SPAdes v3.5.0 genome assembly pipeline. The manual for this version is available [here](http://spades.bioinf.spbau.ru/release3.5.0/manual.html). BayesHammer is a Bayesian subclustering algorithm designed to identify and correct sequencing errors in illumina reads.

_Command Line_  

```
spades.py \
  --only-error-correction \
  -1 SRR2969230_1.fastq \
  -2 SRR2969230_2.fastq \
  -o CMSC702_spades_output		
```

#### Datasets for Assembly
Starting with a single set of RNA-seq reads, *de novo* assemblies were generated for 4 datasets each with a different level of data processing. These processing steps are included in order to improve the quality of the *de novo* assembly. We aim to investigate whether such improvement is detected by the assembly evaluation algorithm.

* __No Modifications__: Raw RNA-seq reads were assembled directly with no further processing.
* __Trimmed Reads__: Raw RNA-seq reads were trimmed with Trim Galore.
* __Corrected Reads__: Error correction was performed on raw RNA-seq reads with BayesHammer.
* __Trimmed & Corrected__: Raw RNA-seq reads were trimmed with Trim Galore and error correction was performed on the trimmed reads with BayesHammer.

#### Assembly

The four differently processed RNA-seq data sets were assembled individually using Trinity v2.0.6 [@Grabherr:2011jb]. The documentation is available [here](https://github.com/trinityrnaseq/trinityrnaseq/wiki). Trinity is a de Bruijn graph based assembly algorithm designed specifically for the *de novo* reconstruction of transcriptomes from RNA-seq reads [@haas2013novo].

_Command Line_
```
Trinity \
  --seqType fq \
  --left SRR2969230_1_val_1.fq \
  --right SRR2969230_2_val_2.fq \
  --CPU $NSLOTS \
  --max_memory 20G \
  --full_cleanup \
  --output Trinity_sequences
```

### Assembly results  
Assembly metrics calculated using Transrate, see http://hibberdlab.com/transrate/metrics.html for a description of the different metrics.  
```{r echo=FALSE, message=FALSE, warning=FALSE}
assembly_metrics_df <- list(no_mods = "data/no_mods/", 
                            corrected = "data/corrected/",
                            trimmed = "data/trimmed/", 
                            trimmedcorrected = "data/trimmedcorrected/") %>% 
      map(paste0, "assemblies.csv") %>%  
      map_df(read_csv, .id = "read_set") %>% 
      select(-assembly) %>% 
      gather("metric","value", -read_set) %>% 
      mutate(read_set = factor(read_set, levels = c("no_mods","corrected","trimmed","trimmedcorrected")))
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
assembly_metrics_df %>% 
      filter(metric %in% c("n_seqs", "smallest", "largest", 
                           "n50", "n_with_orf")) %>%
      spread(metric, value) %>% 
      select(read_set, n_seqs, smallest, largest, n50, n_with_orf) %>% 
      rename(Dataset = read_set, `# of Contigs`=n_seqs, `Shortest Contig (bp)`=smallest,
              `Longest Contig (bp)`=largest, `N50 (bp)`=n50, `# Contigs with ORF`=n_with_orf) %>% 
      kable(digits = 2,
            caption = "Trinity _C. elegans_ transcriptome assembly summary metrics.")
```

N50 is the contig length (bp) where 50% of assembled bases are in contigs of this length or longer.

## Assembly evaluation
### Running TransRate

#### Installation
* Precompiled binaries available from website (http://hibberdlab.com/transrate/), along with instructions to build from source.
* Command to install dependencies `transrate --install-deps type` where type can be either `all`, `read`, or `ref`.
* See Transrate website for additional information for install. 

#### Read Based Evaluation  
For the read based evaluation Transrate maps the raw read data to the transcriptome assembly then uses a set of four metrics to calculate, contig and assembly quality scores. 

__Contig Score:__ the product of the following metrics.  

1. Edit distance - characterization of contig base accuracy, calculated from read-contig edit distance.  
2. Coverage - characterization of based being part of a transcript, calculated from contig based read coverage.  
3. Chimera - probability contig is derived from a single transcript, calculated based on whether the contigs fits a single or double Dirichlet distribution.   
4. Complete - probability contig is complete and correct, calculated using read pair alignments.

__Assembly Score__ is calculated from the mean contig score and proprotion of read pairs mapping to the assembly.  


_Command Line_
```
transrate \
  --left=SRR2969230_1.fastq \
  --right=SRR2969230_2.fastq \
  --assembly=Trinity_sequences.Trinity.fixed.fasta \
  --output=transrate/
```

#### Reference Base Evaluation  
Transrate uses a reciprical blast approach for reference based assembly evaluation. 
A sequence is classified as the reference for a contig if the reference sequence and contig are the top BLAST hits when comparing the reference sequence against all contigs and comparing the contig againts all reference sequences. The reciprical BLAST method is described in @aubry2014deep. 

_Command Line_
```
transrate \
  --reference=Caenorhabditis_elegans.WBcel235.31.pep.all.fa \
  --assembly=Trinity_sequences.Trinity.fixed.fasta \
  --output=transrate/
```

### Transrate results
The unmodified read set was passed as input to Transrate, 
potentally biasing some metric results towards the unmodifed assembly.

#### Read Based Results
##### Contigs
```{r echo=FALSE, message=FALSE, warning=FALSE}
contig_stat <- results_list %>% map(paste0,"contigs.csv") %>% 
      map_df(read_csv, .id = "read_set")

contig_cutoff <- assembly_metrics_df %>% 
      filter(metric == "cutoff") %>% 
      select(-metric) %>% rename(cutoff = value)
contig_stat <- contig_stat %>% left_join(contig_cutoff) %>% 
      mutate(contig_filt = ifelse(cutoff < score,
                                  ">threshold","<threshold"))
```


__Contig Score Distribution__
Trimming reads resulted in a higher proportion of contigs with scores greater than 0.25 then assemblies using unmodified reads or only error corrected reads.  

```{r echo=FALSE, message=FALSE, warning=FALSE}
contig_stat %>% 
      ggplot() + geom_density(aes(x = score, color = read_set, fill = read_set), 
                              alpha = 0.25) + theme_bw()
```

##### Assembly Score
```{r echo=FALSE, message=FALSE, warning=FALSE}
assembly_score_opt <- results_list %>% map(paste0,"assembly_score_optimisation.csv") %>% 
      map_df(read_csv, .id = "read_set") %>% 
      mutate(read_set = factor(read_set, levels = c("no_mods","corrected","trimmed","trimmedcorrected")))
```

Relationship between the cutoff for contig score and assembly score.
```{r echo=FALSE, message=FALSE, warning=FALSE}
assembly_score_opt %>% 
      ggplot() + geom_path(aes(x = cutoff, y = assembly_score, color = read_set)) +
            theme_bw() +
            labs(x = "Contig Score Threshold", y = "Assembly Score")
```


###### Assembly Score Summary
Read trimming and error correction resulted in the highest weighted assembly score, 
while having the lowest contig score cutoff, 
but a lower optimal score.  

Weighted assembly score - takes into consideration expression level, can bias assemblies with high proportions of low expressed poorly assembled contigs.

```{r echo=FALSE, message=FALSE, warning=FALSE}
assembly_metrics_df %>% 
      filter(metric %in% c("score", "optimal_score", "cutoff", "weighted")) %>%
      spread(metric, value) %>% select(read_set, score, optimal_score, cutoff, weighted) %>% 
      rename(Dataset=read_set, `Assembly Score`=score, `Optimized Score`=optimal_score, `Contig Score Threshold`=cutoff, `Weighted Score`=weighted) %>% 
      kable(digits = 3,
            caption = "Trinity _C. elegans_ transcriptome Transrate score summary.")
```



##### Contig Score Breakdown
The following figures show the relationship between individual contig metric distributions and the optimized assembly contig score threshold. 
For all four metrics the metric distribution shifts to the right (increases) for contigs with scores above the optimized assembly contig score threashold.

###### Edit distance 
Mean per-base contig edit distance, average number of base pair differences between contig and reads.

```{r echo=FALSE, message=FALSE, warning=FALSE}
contig_stat %>% 
      ggplot() + geom_histogram(aes(x = p_seq_true)) +
            theme_bw() +
            labs(x = "Mean Per-Base Edit Distance", 
                 y = "Count") +
            facet_grid(contig_filt~read_set) +
            theme(legend.position = "bottom")
```

###### Coverage 
Proportion of bases in the contig covered, at least one read aligned to the contig coverage that position.

```{r echo=FALSE, message=FALSE, warning=FALSE}
contig_stat %>% 
      ggplot() + geom_histogram(aes(x = p_bases_covered)) +
            theme_bw() +
            labs(x = "Proportion of Covered Bases", 
                 y = "Count") +
            facet_grid(contig_filt~read_set) +
            theme(legend.position = "bottom")
```


###### Chimera 
Probability that a contig is not a chimeric misassembly, 
where sequences from two different contigs are incorrectly assembled together.

```{r echo=FALSE, message=FALSE, warning=FALSE}
contig_stat %>% 
      ggplot() + geom_histogram(aes(x = p_not_segmented)) +
            theme_bw() +
            labs(x = "Proportion of Non-Segmented Contigs", 
                 y = "Count") +
            facet_grid(contig_filt~read_set) +
            theme(legend.position = "bottom")
```

###### Complete 
Proportion of bases in reads map to the transcript that support the transcript contig. 
```{r p_good_density, fig.cap = "Histogram of the proportion of bases in agreement with contig and contig length.", echo=FALSE, message=FALSE, warning=FALSE}
contig_stat %>% 
      ggplot() + geom_histogram(aes(x = p_good)) +
            theme_bw() +
            labs(x = "Proportion of Bases Agree with Transcript",
                 y = "Count") +
            facet_grid(contig_filt~read_set) +
            theme(legend.position = "bottom")
```



#### Reference Based Evaluation
Contigs were evaluated based on the reference coverage for contigs with CRB hits, using the _C. elegans_ proteome as the reference. 
Reference coverage was calculated as the proportion of the reference transcript covered by the contig.  

```{r source_ref_dir, warning=FALSE, message=FALSE, echo = FALSE}
results_ref <- list(
      no_mods = "data/no_mods_ref/Trinity_sequences.Trinity.fixed/",
      corrected = "data/corrected_ref/Assembly4_Corrected.Trinity.fixed/",
      trimmed = "data/trimmed_ref/Assembly2_Trimmed.Trinity.fixed/",
      trimmedcorrected = "data/trimmedcorrected_ref/Assembly3_TrimmedCorrected.Trinity.fixed/"
)
```

```{r, message=FALSE, warning=FALSE, echo=FALSE}
ref_contig <- results_ref %>% map(paste0,"contigs.csv") %>% 
      map_df(read_csv, .id = "read_set")
crb_contig <- ref_contig %>% filter(has_crb == "true")
contig_crb_score <- crb_contig %>% left_join(contig_stat)
```

Relationship between reference coverage and contig length.
```{r, message=FALSE, warning=FALSE, echo=FALSE}
ggplot(crb_contig) + geom_hex(aes(x = length, y = reference_coverage)) + 
      facet_wrap(~read_set) + theme_bw() + labs(x = "Contig Length", y = "Reference Coverage")
```

Relationship between reference coverage and contig ORF length.
```{r, message=FALSE, warning=FALSE, echo=FALSE}
ggplot(crb_contig) + geom_hex(aes(x = orf_length, y = reference_coverage)) +
      facet_wrap(~read_set) + theme_bw() + labs(x = "Contig ORF Length", y = "Reference Coverage")
```


#### Read and Reference Evaluation Comparison  
No clear relationship between contig score and reference coverage.  

```{r, message=FALSE, warning=FALSE, echo=FALSE, fig.cap = "Relationship between reference and read based quality assessment.", fig.width = 12}
ggplot(contig_crb_score) + 
      geom_density2d(aes(x = reference_coverage, y = score), color = "darkblue") +
      geom_smooth(aes(x = reference_coverage, y = score), color = "darkorange") +
      facet_grid(contig_filt~read_set) +
      theme_bw() + labs(x = "Reference Transcript Coverage", y = "Contig Score")
```

## Conclusions
- Read trimming has a greater impact on assembly size and number of contigs than error correction.  
- Our read based assembly evaluation potentially biased by the use of un-modified reads to evalulate all four assemblies. The trimmed read assemblies might be penalized as some of the untrimmed reads were unable to align the to the trimmed read assemblies.  
- We observed no clear relationship between contig score and reference coverage, this is potentially due to correct contigs shorter than the reference transcript having low reference coverage.  

## Session Information
__Session Info__
```{r s_info, echo=FALSE, message=FALSE, warning=FALSE}
s_info <- devtools::session_info()
print(s_info$platform)
kable(s_info$packages)
```

## References
